### PR TITLE
Unified testing framework

### PR DESCRIPTION
**Pull request by HebaruSan**
## Problem

See #210, installing mypy in the Dockerfile caused problems related to dependencies-of-dependencies that were installed at the system level and then not available on a fresh image with only the user install copied over.

## Motivation

In the Discord discussion of the first iteration of this pull request, we agreed that it would be nice to have one simple command that developers could run to test everything, and to re-use that same command in our GitHub workflows. It looked like the `test_suite` option in `setup.py`, supporting the `python setup.py test` command, might be a step in that direction.

Enter pytest-dev/pytest#5534. They're deprecating `python setup.py test`.

In the Discord discussion of the _second_ iteration of this pull request, `tox` was ruled out as a solution because we already use Docker to manage our virtual environments and would prefer to run our tests in the same environment where the actual code runs. So now `pytest` is winning.

## Changes

- Now we don't install or run mypy in the Dockerfile
- A new `test` workflow is created to run all tests
- The separate mypy and unittest workflows are deleted

For me, this is how to run tests, though if you use a distro that doesn't try to mix Python 2 and 3 like Ubuntu does, drop the `-3`:

```
cd netkan
pip install --user .[test]
pytest-3 --mypy -v
```

This provides a single uniform access point for all testing, which will work the same for developers and for CI/CD.

### CODE REVIEW

**Comment by HebaruSan** [comment]
I guess unittest is not a package?

**Comment by techman83** [comment]
> I guess unittest is not a package?

`unittest` is a part of python standard lib.

File: .github/workflows/mypy.yml [hunk 1]
```diff
@@ -14,8 +16,8 @@ jobs:
         python-version: 3.7
     - name: Install Dependencies
       run: |
-        pip install netkan/.[test]
+        pip install --user netkan/. --no-warn-script-location --force-reinstall
+        pip install mypy
```
Comment (techman83) [inline .github/workflows/mypy.yml:20]: Our test dependencies *should* come from within `setup.py`. The reason pip was installing mypy directly was that calling `[test]` will install all the deps plus the test ones. `--force-reinstall` ensures all deps get installed as user, there isn't any need to install as user in this environment.

File: .github/workflows/unittest.yml [hunk 2]
```diff
@@ -14,8 +16,7 @@ jobs:
         python-version: 3.7
     - name: Install Dependencies
       run: |
-        pip install netkan/.[test]
+        pip install --user netkan/. --no-warn-script-location --force-reinstall
     - name: Run unittest
```
Comment (techman83) [inline .github/workflows/unittest.yml:20]: See previous comment.

File: netkan/Dockerfile [hunk 3]
```diff
@@ -3,12 +3,9 @@ RUN useradd -ms /bin/bash netkan
 ADD . /netkan
 WORKDIR /netkan
 RUN pip install pip --upgrade
-RUN pip install mypy
 RUN chown -R netkan:netkan /netkan
 USER netkan
 RUN pip install --user . --no-warn-script-location --force-reinstall
-RUN mypy .
-RUN python -m unittest -v
```
Comment (techman83) [inline netkan/Dockerfile:9]: Our docker environment isn't our github one, we likely still want to run our tests inside the container.

File: netkan/setup.py [hunk 4]
```diff
@@ -38,9 +38,6 @@
             'pylint',
             'autopep8',
             'troposphere',
-        ],
-        'test': [
-            'mypy',
         ]
```
Comment (techman83) [inline netkan/setup.py:41]: Our package should be explicit about its test dependencies.

**Review by techman83 — COMMENTED** [review]
I think we can drop `mypy` from the container, it's non-runtime type checking. But our tests should still be run inside the container.

Though we don't use `pytest` or any other test packages in the future, so we may want to bust out `mypy` into a separate dependency set.

File: .github/workflows/mypy.yml (hunk 1 above)
Comment (HebaruSan) [reply .github/workflows/mypy.yml:20]: I don't think we've been well-served by the concept of "test dependencies". We need mypy here and nowhere else, and installing it elsewhere caused 10 hours of downtime.

File: netkan/Dockerfile (hunk 3 above)
Comment (HebaruSan) [reply netkan/Dockerfile:10]: Makes sense to me.

File: netkan/setup.py (hunk 4 above)
Comment (HebaruSan) [reply netkan/setup.py:41]: Mypy isn't needed to run unittest. I think we should handle such things individually, so we only have the tools we need in each case.

Comment (techman83) [reply netkan/setup.py:41]: `mypy` is technically a test dependency, but you are correct it isn't a `unittest` dependency. However if we say move to `pytest`, we would actually have more than one dependency for testing.

Comment (HebaruSan) [reply netkan/setup.py:41]: But they would still be separate dependencies for separate workflows. What's the benefit in combining them into one big list and installing extra stuff?

File: .github/workflows/mypy.yml (hunk 1 above)
Comment (techman83) [reply .github/workflows/mypy.yml:20]: Well, it was the manual installation of `mypy` in the container that caused the downtime. We won't be installing our test dependencies in the container for now and if we do, we'll need to think about how best to approach it. For example we could have a separate 'test' stage, that is `FROM production as test` that we could run as an build action

```Dockerfile
FROM production as test
USER root
ADD . /netkan
RUN chown -R netkan:netkan /netkan
WORKDIR /netkan
USER netkan
RUN pip install --user /netkan/.[test]
RUN mypy .
RUN pytest -v
```


```yml
      - name: Build Test
        id: docker_build
        uses: docker/build-push-action@v2
        with:
          context: ./netkan
          file: ./netkan/Dockerfile
          push: false
          target: test
```

File: netkan/setup.py (hunk 4 above)
Comment (techman83) [reply netkan/setup.py:41]: The benefit is that if you are someone who wants to run our tests, that you get all the dependencies required for running said tests. What we do in CI/CD we can document in code; but if a python dev comes along with a fix, they should be able to go `pip install .[test]` and have everything they require to run them.

It's more about our package being correct than being helpful for ci. Though we could add a separate stanza for ci only packages if we want to be clear about it.

File: .github/workflows/unittest.yml (hunk 2 above)
Comment (techman83) [reply .github/workflows/unittest.yml:19]: We actually don't use any dependencies for `unittest` - we can adjust that later if that changes.

File: netkan/setup.py (hunk 4 above)
Comment (HebaruSan) [reply netkan/setup.py:41]: Hmm, mypy isn't a test, though, in the sense you're talking about.

**Comment by HebaruSan** [comment]
It's not running the old workflows or the new one. Clearly I've missed something. Any tips appreciated.

File: netkan/setup.py (hunk 4 above)
Comment (HebaruSan) [reply netkan/setup.py:41]: The new `tox.ini` file specifies the test dependencies in the `deps =` line.

**Review by techman83 — COMMENTED** [review]
Tox has its use case, but I feel it's a little redundant. We can already manage environments in our workflows and our container.

`pytest` is the go to for testing frameworks. It has wrappers for mypy, linting, flake8 etc. Which would allow us to close off #49

It is rather daft that the `python setup.py test` is now deprecated, but the python community have been re-inventing the packaging wheel for as long as I can remember.

**Comment by techman83** [comment]
Ok, so I've put my code where my mouth is! :slightly_smiling_face: 

* `test` stage for docker, which uses production as a base. These layers won't end up in production as we target `production`, but will test in the same environment.
* keeping the `--help` check in production is still a good sanity check
* pytest used in the test workflow
* tests are run before the deployment is produced
* vs code environment works correctly (this was a right pain..)

**Comment by HebaruSan** [comment]
What's the command for running tests locally now? I'm not finding one that works yet:

```
$ pytest
Command 'pytest' not found, did you mean:
  command 'ptest' from deb pbuilder-scripts (22)
  command 'ptest' from deb parmetis-test (4.0.3-5build1)
  command 'yptest' from deb nis (3.17.1-5)
Try: sudo apt install <deb name>

$ py.test-3 --mypy -v
ERROR: usage: py.test-3 [options] [file_or_dir] [file_or_dir] [...]
py.test-3: error: unrecognized arguments: --mypy
  inifile: /home/user/Downloads/KSP/NetKAN-Infra/netkan/pytest.ini
  rootdir: /home/user/Downloads/KSP/NetKAN-Infra/netkan
```

**Comment by HebaruSan** [comment]
@techman83 your changes look good to me. Should we look into pylint and flake8?

**Comment by techman83** [comment]
Yes, but I think a separate PR. I enabled pylint and got a whole suite of warnings I didn't want to deal with on a PR for unifying our testing.

**Pull request merged**